interop/ksp/src/main/kotlin/com/squareup/kotlinpoet/ksp/originatingKSFiles.kt

```diff
@@ -0,0 +1,169 @@
+}
+
+/** Returns this spec's originating [KSFiles][KSFile] for use with incremental processing. */
+@KotlinPoetKspPreview
+public fun TypeSpec.originatingKSFiles(): List<KSFile> = getKSFilesTag()
+
+/** Returns this spec's originating [KSFiles][KSFile] for use with incremental processing. */
+@KotlinPoetKspPreview
+public fun FunSpec.originatingKSFiles(): List<KSFile> = getKSFilesTag()
+
+/** Returns this spec's originating [KSFiles][KSFile] for use with incremental processing. */
+@KotlinPoetKspPreview
+public fun PropertySpec.originatingKSFiles(): List<KSFile> = getKSFilesTag()
+
+/** Returns this spec's originating [KSFiles][KSFile] for use with incremental processing. */
+@KotlinPoetKspPreview
+public fun TypeAliasSpec.originatingKSFiles(): List<KSFile> = getKSFilesTag()
+
+/**
+ * Returns the list of all files added to the contained
+ * [TypeSpecs][TypeSpec], [PropertySpecs][PropertySpec], [FunSpecs][FunSpec], or
+ * [TypeAliasSpecs][TypeAliasSpec] contained in this spec.
+ */
+@KotlinPoetKspPreview
+public fun FileSpec.originatingKSFiles(): List<KSFile> {
+  return members
+    .flatMap {
+      when (it) {
+        is FunSpec -> it.originatingKSFiles()
+        is PropertySpec -> it.originatingKSFiles()
+        is TypeSpec -> it.originatingKSFiles()
+        is TypeAliasSpec -> it.originatingKSFiles()
+        else -> emptyList()
+      }
+    }
+    .distinct()
+}
+
+/** Adds the given [ksFile] to this builder's tags for use with [originatingKSFiles]. */
+@KotlinPoetKspPreview
+public fun TypeAliasSpec.Builder.addOriginatingKSFile(ksFile: KSFile): TypeAliasSpec.Builder = apply {
+  getOrCreateKSFilesTag().add(ksFile)
+}
+
+/** Adds the given [ksFile] to this builder's tags for use with [originatingKSFiles]. */
+@KotlinPoetKspPreview
+public fun PropertySpec.Builder.addOriginatingKSFile(ksFile: KSFile): PropertySpec.Builder = apply {
+  getOrCreateKSFilesTag().add(ksFile)
+}
+
+/** Adds the given [ksFile] to this builder's tags for use with [originatingKSFiles]. */
+@KotlinPoetKspPreview
+public fun FunSpec.Builder.addOriginatingKSFile(ksFile: KSFile): FunSpec.Builder = apply {
+  getOrCreateKSFilesTag().add(ksFile)
+}
+
+/** Adds the given [ksFile] to this builder's tags for use with [originatingKSFiles]. */
+@KotlinPoetKspPreview
+public fun TypeSpec.Builder.addOriginatingKSFile(ksFile: KSFile): TypeSpec.Builder = apply {
+  getOrCreateKSFilesTag().add(ksFile)
+}
+
+/**
+ * Writes this [FileSpec] to a given [codeGenerator] with the given [originatingKSFiles].
+ *
+ * Note that if none are specified, the [originatingKSFiles] argument defaults to using
+ * [FileSpec.originatingKSFiles], which will automatically resolve any files added to the
+ * contained declarations.
+ *
+ * See [the docs](https://github.com/google/ksp/blob/main/docs/incremental.md) for more information.
+ *
+ * @see FileSpec.originatingKSFiles
+ * @param codeGenerator the [CodeGenerator] to write to
+ * @param aggregating flag indicating if this is an aggregating
```

Comment:
nit: "aggregating annotation processor"?

Reply:
Good catch, though I think we want to say "symbol processor"

Reply:
Agree

Reply:
cb6f2c5